### PR TITLE
Maya: Workfile testing

### PR DESCRIPTION
**Pull request by tokejepsen**
# TODO
- [x] The menu entry needs to be hidden from artists, behind some environment variable.

## Changelog Description
This adds testing code to run in Maya for regression testing.

This is not meant to be comprehensive, but merely a starting point for making tests. The testing does not touch on areas like launching and cross-DCC testing, so the database etc. is meant to made manually for now. This also means it does not overlap very much with existing OpenPype testing framework.

The PR adds a `Testing` menu entry to the `AYON` menu in Maya:

Screenshot 2024-03-26 113055

The options are are split into 2 categories; `Repository Workfile` and `Current Workfile`.

`Repository Workfile` - will take the workfile from the repository, save to the current context as a new workfile version and run tests on it.
`Current Workfile` - will run tests on the current workfile.

This is will help with iteration of the workfile in the repository that should be a ground truth for a functional pipeline. Users can run the tests on the repository workfile initially, then continue iterating on the current workfile, before save to the repository again for committing.

In each category there are 4 options:

- `Run Tests...` which will run through the whole testing suite.
- `Test Create...` which will test the creation stage, by re-creating all instances in the workfile.
- `Test Publish...` which will run the validation recursively until its either successful or there are no more approved actions to execute, like repair actions. If validation is successful, it'll run a full publish.
- `Test Load...` which will run the template builder on the placeholders in the workfile.

I've run through these tests while testing various PRs, which yielded a couple of bugs that only makes sense to fix in this PR as well:

- when creating a new `render` instance with Arnold, the settings gets completely cleared to default with `unifiedRenderGlobalsRevertToDefault`, which means we'll need to restore the camera used at least to have a reproducible creation stage.
- since the default state of AOVs in Arnold is to merge, we modify the prefix to not have `RenderPass` token.
- when creating instances, the code did not return the instances created.
- when creating placeholders with boolean values for arguments, the naming of the placeholder node would error.
- when populating the placeholders the placeholder set would not be created.
- when ordering placeholders account for locked nodes.
- removing a `rig` instance would not completely remove all sets.
- enabling file node loading for reviews.
- when populating the placeholders there was no return value. Returning success and the placeholders produced.
- split camera token validation since the ValidateRenderSettings would offer repair action for users without actually repairing the issue. This also helps with the recursive validation.

## Testing notes:
1. Create Maya addon package and upload to server.
2. Enable `ayon+settings://maya/workfile_testing?project=demo_Commercial`
3. Launch Maya and run `Run Tests On Repository Workfile`.

### CODE REVIEW

**Comment by tokejepsen** [comment]
@BigRoy this is now behind a project setting.

**Comment by moonyuet** [comment]
i tested with `Run Tests On Repository Workfile` and all testing functions work as expected. 
I also test with `Run Tests On current Workfile`
image
image
It will hit Type Error when there is no placeholder found,

**Comment by tokejepsen** [comment]
@moonyuet thanks for testing. The issue with no placeholders is accounted for this with an assertion message that'll inform user that we'll need placeholders to test the loading.

**Comment by moonyuet** [comment]
> @moonyuet thanks for testing. The issue with no placeholders is accounted for this with an assertion message that'll inform user that we'll need placeholders to test the loading.

I did the test again, all works as expected with the sample scene and my own scene. It raises the error when testing to load objects when there is no placeholder as expected.

**Review by moonyuet — APPROVED** [review]

**Comment by tokejepsen** [comment]
Waiting on https://github.com/ynput/ayon-core/pull/478 to be merged before fixing the tests.

**Review by moonyuet — CHANGES_REQUESTED** [review]
I encountered some errors when I click `Test Publish on Repository Workfile`
```# Traceback (most recent call last):
# #   File "D:\ayon-core\client\ayon_core\hosts\maya\api\testing\lib.py", line 97, in test_publish_on_repository_workfile
    test_publish()
# #   File "D:\ayon-core\client\ayon_core\hosts\maya\api\testing\tests.py", line 182, in test_publish
    success, context = recursive_validate(
# #   File "D:\ayon-core\client\ayon_core\hosts\maya\api\testing\tests.py", line 139, in recursive_validate
    action().process(context, result["plugin"])
# #   File "D:\ayon-core\client\ayon_core\pipeline\publish\publish_plugins.py", line 243, in process
    plugin.repair(instance)
#   File "<string>", line 100, in repair
#   File "<string>", line 40, in get_invalid_resolution
#   File "<string>", line 87, in get_folder_resolution
# KeyError: 'taskEntity'
# Error: KeyError: file <string> line 87: 'taskEntity'
```

**Comment by tokejepsen** [comment]
> I encountered some errors when I click Test Publish on Repository Workfile

@moonyuet strange, I cant replicate this, so its probably DCC context issues. I've improved the error messages on actions. Could you try again?

**Comment by moonyuet** [comment]
> > I encountered some errors when I click Test Publish on Repository Workfile
> 
> @moonyuet strange, I cant replicate this, so its probably DCC context issues. I've improved the error messages on actions. Could you try again?

It's still not working with `Test Publish on Repository Workfile` (and `Test Load on Repository Workfile` too). The attachment are the log messages for the testing.
[Test_Load_Repository_Workfile.txt](https://github.com/ynput/ayon-core/files/15272401/Test_Load_Repository_Workfile.txt)
[Test_Publish_Repository_Workfile.txt](https://github.com/ynput/ayon-core/files/15272403/Test_Publish_Repository_Workfile.txt)

And also not sure if it is expected behavior to have this message showed in terms of OCIO config(It can be other PRs but just wondering as I saw your name in the message)
```
# Warning: Error could not read 'C:/Users/tokejepsen/AppData/Local/Ynput/AYON/addons/ayon_ocio_1.0.0/ayon_ocio/configs/OpenColorIOConfigs/aces_1.2/config.ocio' OCIO profile.
// Warning: line 1233: The default image may not be modified. Use the -i/image flag instead.
```

**Comment by moonyuet** [comment]
@tokejepsen I have made some adjustment on the task entity and folder entity so that the testing script won't hit any KeyError when testing publish.
@antirotor @BigRoy can you find some time to look at this?

File: client/ayon_core/hosts/maya/api/action.py [hunk 1]
```diff
@@ -90,7 +90,10 @@ def _update_id_attribute(self, instance, nodes):
             folder_entity = ayon_api.get_folder_by_path(
                 project_name, folder_path, fields={"id"}
             )
-
+            if not folder_entity:
+                self.log.debug(
+                    f"No folder entity found in the project {project_name}")
+                return
```
Comment (BigRoy) [inline client/ayon_core/hosts/maya/api/action.py:96]: When would this bug occur? This seems like a very dangerous point to get to regardless of the action. An instance with an invalid folder path is not allowed in the new publisher by design - right @iLLiCiTiT ?

Aside of that, if we do need this for whatever reason I'd say this should also log the folder path it tried to find.

Something like: `Folder {folder_path} not found in project {project_name}`

File: client/ayon_core/hosts/maya/plugins/create/create_rig.py [hunk 2]
```diff
@@ -20,13 +26,28 @@ def create(self, product_name, instance_data, pre_create_data):
...
+        for node, id in lib.generate_ids(sets):
+            lib.set_id(node, id, overwrite=True)
+
+        return instance
+
+    def remove_instances(self, instances):
+        for instance in instances:
+            nodes = [instance.data.get("instance_node")]
+            for suffix in self.set_suffixes:
+                nodes.append(instance.data.get("instance_node") + suffix)
```
Comment (BigRoy) [inline client/ayon_core/hosts/maya/plugins/create/create_rig.py:45]: Wouldn't it just make more sense to check if the instance node contains any of the sets as members, and only remove them then? Just going by set name and suffix seems like prone to bugs?

File: client/ayon_core/hosts/maya/plugins/workfile_build/load_placeholder.py [hunk 3]
```diff
@@ -102,6 +102,11 @@ def _parent_in_hierarchy(self, placeholder, container):
             container_root = parent
         roots = [container_root]
 
+        # Some loaded containers do not produce any DAG nodes, for example
+        # rendersetup.
+        if not container_root:
+            return
```
Comment (BigRoy) [inline client/ayon_core/hosts/maya/plugins/workfile_build/load_placeholder.py:108]: This should technically be in a different PR.

File: client/ayon_core/hosts/maya/api/plugin.py [hunk 4]
```diff
@@ -484,6 +485,9 @@ def collect_instances(self):
 
             instance.transient_data["layer"] = layer
             self._add_instance_to_context(instance)
+            instances.append(instance)
+
+        return instances
```
[the remark this answers is not in the document]
Comment (tokejepsen) [reply client/ayon_core/hosts/maya/api/plugin.py:490]: We need it to get the instances when re-creating them. Might not be a requirement, but thought it was/should be common practice of the plugins to return the generated instances?

File: client/ayon_core/hosts/maya/api/testing/lib.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    workfile_path = os.path.join(workdir, workfile)
+
+    host = registered_host()
+    host.open_file(os.path.join(os.path.dirname(__file__), "tests.ma"))
+    host.save_file(workfile_path)
+
+
+def recursive_validate(valid_action_names):
+    """ Recursively validate until until it is either successfull or there are
+    no more approved actions to run in which case its failing.
```
[the remark this answers is not in the document]
Comment (tokejepsen) [reply client/ayon_core/hosts/maya/api/testing/lib.py:70]: > Looking at the logic here this will infinitely recurse if the action does run without a 'hard error' in the logic but actually does NOT fix the validation. Right? Similarly - there may be some cases where a validation repair of one action, may trigger another validation which's repair may trigger the original plug-in's validation again infinitely recursing. We should keep track of which plug-in's actions were triggered and disallow those from being triggered again since that should actually be considered a bug in itself.

True, I've had infinitely recursion before, so any ways of accounting for this would be great!

File: client/ayon_core/hosts/maya/api/testing/tests.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+from . import lib
+
+
+# Needed for transition phase for asset/subset renaming. Can be hardcoded once
+# transition is done.
+product_key_name = "productName"
+product_type_key_name = "productType"
+
+
+def test_create():
```
[the remark this answers is not in the document]
Comment (tokejepsen) [reply client/ayon_core/hosts/maya/api/testing/tests.py:27]: Not entirely sure what you mean, but we are going through the creators `create` method.

We are querying the members of the sets, since the selection when creating the instances are not recording elsewhere.

File: client/ayon_core/hosts/maya/api/plugin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BigRoy) [reply client/ayon_core/hosts/maya/api/plugin.py:490]: Wouldn't the `CreateContext` contain the collected instances? We shouldn't be calling `Creator.collect_instances()` directly, right?

File: client/ayon_core/hosts/maya/api/testing/tests.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    context = pyblish.util.publish()
+    success = True
+    for result in context.data["results"]:
+        if not result["success"]:
+            success = False
+            break
+
+    assert success, lib.create_error_report(context)
+
+    print("Publish was successfull!")
```
Comment (BigRoy) [inline client/ayon_core/hosts/maya/api/testing/tests.py:136]:
```suggestion
    print("Publish was successful!")
```

File: client/ayon_core/hosts/maya/api/testing/tests.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                    type(err),
+                    err,
+                    err.__traceback__
+                )
+            )
+            error_message += "\n" + formatted_traceback
+
+    assert success, error_message
+
+    print("Load was successfull!")
```
Comment (BigRoy) [inline client/ayon_core/hosts/maya/api/testing/tests.py:197]:
```suggestion
    print("Load was successful!")
```

File: client/ayon_core/hosts/maya/api/testing/tests.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BigRoy) [reply client/ayon_core/hosts/maya/api/testing/tests.py:27]: I wonder why we're testing "recreating" instances? Shouldn't we just check whether a regular "create" works?

[the remark this answers is not in the document]
Comment (BigRoy) [reply client/ayon_core/hosts/maya/api/testing/tests.py:27]: Anyway, I meant going through `CreateContext.create` instead of `Creator.create`

Like:
```python
created_instance = context.create(
    creator_identifier=creator_identifier,
    variant=variant,
    pre_create_data={"use_selection": True}
)
```

Similarly, I would set the creator attributes and publish attribute values via the created instance:

```python
for key, value in old_instance.creator_attributes:
    created_instance.creator_attributes[key] = value
```

And then allowing the CreateContext and `update_instances` logic of the `Creators` to trigger and ensure the saving of the instances logic works as intended.

```python
context.save_changes()
```

Making something like:

```python
created_instance = context.create(
    creator_identifier=creator_identifier,
    variant=variant,
    pre_create_data={"use_selection": True}
)

for key, value in old_instance.creator_attributes:
    created_instance.creator_attributes[key] = value

context.save_changes()
```

---

Also, pretty sure we want `use_selection` to be `False` since we apparently seem to be forcing members into the created instances to match the original instance's members?

[the remark this answers is not in the document]
Comment (tokejepsen) [reply client/ayon_core/hosts/maya/api/testing/tests.py:27]: Cool, this all sounds good.

File: client/ayon_core/hosts/maya/api/action.py (hunk 1 above)
Comment (iLLiCiTiT) [reply client/ayon_core/hosts/maya/api/action.py:96]: > When would this bug occur?

Same question?

If it can somehow happen (which should not), shouldn't the action actually fail in that case as it seems like "all fixed" but on next publish it will fail again?

File: client/ayon_core/hosts/maya/api/plugin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply client/ayon_core/hosts/maya/api/plugin.py:490]: > We shouldn't be calling Creator.collect_instances() directly, right?

That should not happen indeed.

File: client/ayon_core/hosts/maya/plugins/publish/validate_node_ids_related.py [hunk 9]
```diff
@@ -61,11 +61,14 @@ def process(self, instance):
     @classmethod
     def get_invalid(cls, instance):
         """Return the member nodes that are invalid"""
+        invalid = list()
+        if not instance.data.get("folderEntity"):
+            return invalid
```
[the remark this answers is not in the document]
Comment (moonyuet) [reply client/ayon_core/hosts/maya/plugins/publish/validate_node_ids_related.py]: when creating the test and publish, it always shows up the TypeError where the folderEntity or taskEntity is not found.

[the remark this answers is not in the document]
Comment (BigRoy) [reply client/ayon_core/hosts/maya/plugins/publish/validate_node_ids_related.py]: Do you have the Stack Trace?

If it's a `TypeError` then at least it has found an entry in there - but it's of the wrong type? How?
